lib/steep/type_construction.rb

```diff
@@ -1813,7 +1813,7 @@ def synthesize(node, hint: nil, condition: false)
                 .for_branch(right_node)
                 .synthesize(right_node, hint: left_truthy.type, condition: true).to_ary
 
-            right_truthy, right_falsy = interpreter.eval(env: left_falsy.env, node: right_node)
+            right_truthy, right_falsy = interpreter.eval(env: right_context.type_env, node: right_node)
```

Comment:
We need `right_context.type_env` instead of `left_falsy.env`.
The `right_context` has purities of method calls in addition to variable typings in `left_falsy.env`.